[wabt] Add new port

alexreinking opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] The name of the port matches an existing name for this component on https://repology.org/ if possible, and/or is strongly associated with that component on search engines.
- [x] Optional dependencies are resolved in exactly one way. For example, if the component is built with CMake, all `find_package` calls are REQUIRED, are satisfied by `vcpkg.json`'s declared dependencies, or disabled with [CMAKE_DISABLE_FIND_PACKAGE_Xxx](https://cmake.org/cmake/help/latest/variable/CMAKE_DISABLE_FIND_PACKAGE_PackageName.html).
- [x] The versioning scheme in `vcpkg.json` matches what upstream says.
- [x] The license declaration in `vcpkg.json` matches what upstream says.
- [x] The installed as the "copyright" file matches what upstream says.
- [x] The source code of the component installed comes from an authoritative source.
- [x] The generated "usage text" is accurate. See [adding-usage](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/examples/adding-usage.md) for context.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is in the new port's versions file.
- [x] Only one version is added to each modified port's versions file.

alexreinking committed:
Add wabt to vcpkg

MonicaLiu0311 commented on ports/wabt/portfile.cmake:
```diff
@@ -0,0 +1,43 @@
+)
+
+vcpkg_cmake_install()
+vcpkg_cmake_config_fixup()
+
```
```suggestion
vcpkg_cmake_config_fixup()
if("tools" IN_LIST FEATURES)
    vcpkg_copy_tools(
        TOOL_NAMES spectest-interp wasm2c wasm2wat wasm-decompile wasm-interp wasm-objdump wasm-stats wasm-strip wasm-validate wast2json wat2wasm wat-desugar
        AUTO_CLEAN)
endif()

```

MonicaLiu0311 replied:
Solve post-build errors:
![image](https://github.com/user-attachments/assets/7d9fb630-1098-4ae7-a41c-0b078eece00f)

alexreinking replied:
Done!

alexreinking replied:
Interestingly, I did not observe those post-build errors on macOS. I wonder if those validations only happen or work on Windows?

alexreinking replied:
Actually, I thought it would be easier to maintain the port if the `OPTIONS_RELEASE` are set to install the tools to the correct directory in the first place.

MonicaLiu0311 commented on ports/wabt/portfile.cmake:
```diff
@@ -0,0 +1,43 @@
+vcpkg_check_linkage(ONLY_STATIC_LIBRARY)
```
Could you please explain why only static builds are supported?

alexreinking replied:
The library is marked `STATIC` in the upstream build files and does not attempt to export symbols.

MonicaLiu0311 commented on ports/wabt/vcpkg.json:
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "wabt",
+  "version": "1.0.36",
```
I didn't find any version information upstream: https://github.com/WebAssembly/website/tags.

alexreinking replied:
That's the wrong repository. See here:

https://github.com/WebAssembly/wabt/tags

MonicaLiu0311 commented:
All features are tested successfully in the following triplet:
```
x86-windows
x64-windows
x64-windows-static
```

The usage test passed on `x64-windows` (header files found):
```
wabt provides CMake targets:

  # this is heuristically generated, and may not be correct
  find_package(wabt CONFIG REQUIRED)
  target_link_libraries(main PRIVATE wabt::wabt wabt::wasm-rt-impl)
```

alexreinking commented:
Thank you, @MonicaLiu0311 !

dg0yt commented on ports/wabt/portfile.cmake:
```diff
@@ -0,0 +1,47 @@
+
+# wabt enables wasm-rt-impl iff setjmp.h is found by `check_include_file`.
+# It does not use this variable otherwise.
+vcpkg_check_features(OUT_FEATURE_OPTIONS wasm_rt_impl_options
+                     FEATURES wasm-rt-impl HAVE_SETJMP_H)
+
+vcpkg_check_features(OUT_FEATURE_OPTIONS tools_options
+                     FEATURES tools BUILD_TOOLS)
```
The split is difficult to understand in future updates. If possible, it should be done in one call IMO.

alexreinking replied:
This is done to avoid building the tools in debug mode since they aren't installed anyway.

dg0yt commented on ports/wabt/portfile.cmake:
```diff
@@ -0,0 +1,47 @@
+
+# wabt tries to vendor picosha2. Since it's a single-header library, we can
+# simply inject its location into the standard include directories.
+set(picosha2 -DCMAKE_CXX_STANDARD_INCLUDE_DIRECTORIES=${CURRENT_INSTALLED_DIR}/include)
```
```suggestion
set(picosha2 "-DCMAKE_CXX_STANDARD_INCLUDE_DIRECTORIES=${CURRENT_INSTALLED_DIR}/include")
```

dg0yt replied:
Actually I think this doesn't belong to `CMAKE_CXX_STANDARD_INCLUDE_DIRECTORIES`. This variables belongs to the toolchain, not to the users.

alexreinking replied:
So you would prefer a patch?

dg0yt replied:
It doesn't have to be a patch. It could be a CMAKE_PROJECT_INCLUDE with a regular `include_directories("${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include")` (or using `find_path` first).

dg0yt commented on ports/wabt/portfile.cmake:
```diff
@@ -0,0 +1,47 @@
+    OPTIONS_DEBUG
+        -DBUILD_TOOLS=OFF
+    OPTIONS_RELEASE
+        -DCMAKE_INSTALL_BINDIR=tools/wabt
+        ${tools_options}
```
Both can lines be in the regular `OPTIONS`.

alexreinking replied:
No, they cannot be because vcpkg_cmake_configure overwrites them.

dg0yt replied:
The second line will be overwritten by OPTIONS_DEBUG. Which is what you want.

The first line, yeah, vcpkg really has a different idea what should be the value. I understand that you do it to avoid needing `vcpkg_copy_tools`. However it is tightly coupling to the fact that the port doesn't build runtime libs (ATM) and not really good practice IMO.

dg0yt commented on ports/wabt/portfile.cmake:
```diff
@@ -0,0 +1,47 @@
+        ${wasm_rt_impl_options}
+        -DBUILD_LIBWASM=OFF
+        -DBUILD_TESTS=OFF
+        -DCMAKE_POSITION_INDEPENDENT_CODE=ON
```
Toolchain setting, and enabled in all vcpkg toolchains. Not to be set per port.
```suggestion
```

alexreinking replied:
Thanks, will remove.

dg0yt reviewed (commented):
Is the lib actually expected to be used outside the package's own tools?
(Is this more a tool port than a lib?)

alexreinking commented:
> Is the lib actually expected to be used outside the package's own tools?
> (Is this more a tool port than a lib?)

It is an optional dependency of Halide, which is why I'm adding a port here rather than maintaining a separate overlay. The maintainers of wabt are aware of this usage.

alexreinking committed:
version database

alexreinking commented:
Thank you for your review @dg0yt. In response, I have:

1. Removed usage of `CMAKE_POSITION_INDEPENDENT_CODE`
2. Replaced `CMAKE_CXX_STANDARD_INCLUDE_DIRECTORIES` with a file that calls `find_path`+`include_directories` and is injected via `CMAKE_PROJECT_INCLUDES`, as you suggested.
3. Replaced the dubious-for-vcpkg usage of `CMAKE_INSTALL_BINDIR` with `vcpkg_copy_tools`.
4. Merged the two `vcpkg_check_features` calls

I hope this brings us closer 🙂

alexreinking commented:
@dg0yt @MonicaLiu0311 -- what's the process from here?
